File: likecoin-migration/apis/models/likeCoinMigration.ts
import { z } from 'zod';

import { CoinStringSchema } from './coin';

export const LikeCoinMigrationStatusSchema = z.enum([
  'pending_cosmos_tx_hash',
  'verifying_cosmos_tx',
  'evm_minting',
  'evm_verifying',
  'completed',
  'failed',
]);

export type LikeCoinMigrationStatus = z.infer<
  typeof LikeCoinMigrationStatusSchema
>;

export const LikeCoinMigrationSchema = z.object({
  id: z.number(),
  created_at: z.coerce.date(),
  user_cosmos_address: z.string(),
  user_eth_address: z.string(),
  evm_signature: z.string(),
  amount: CoinStringSchema,
  status: LikeCoinMigrationStatusSchema,
  cosmos_tx_hash: z.string().nullable(),
  evm_tx_hash: z.string().nullable(),
  failed_reason: z.string().nullable(),
});

export type LikeCoinMigration = z.output<typeof LikeCoinMigrationSchema>;

export type Pending<M extends LikeCoinMigration> = M & {
  status: 'pending_cosmos_tx_hash';
};

export type Polling<M extends LikeCoinMigration> = M & {
  status: 'verifying_cosmos_tx' | 'evm_minting' | 'evm_verifying';
};

export type Completed<M extends LikeCoinMigration> = M & {
  status: 'completed';
};

export type Failed<M extends LikeCoinMigration> = M & {
  status: 'failed';
  failed_reason: string;
};

File: migration-backend/pkg/handler/model/likecoin_migration.go
package model

import (
	"time"

	"github.com/likecoin/like-migration-backend/pkg/model"
)

type LikeCoinMigration struct {
	Id                uint64    `json:"id"`
	CreatedAt         time.Time `json:"created_at"`
	UserCosmosAddress string    `json:"user_cosmos_address"`
	UserEthAddress    string    `json:"user_eth_address"`
	EvmSignature      string    `json:"evm_signature"`
	Amount            string    `json:"amount"`

	Status       model.LikeCoinMigrationStatus `json:"status"`
	CosmosTxHash *string                       `json:"cosmos_tx_hash"`
	EvmTxHash    *string                       `json:"evm_tx_hash"`
	FailedReason *string                       `json:"failed_reason"`
}

func LikeCoinMigrationFromModel(c *model.LikeCoinMigration) *LikeCoinMigration {
	return &LikeCoinMigration{
		Id:                c.Id,
		CreatedAt:         c.CreatedAt,
		UserEthAddress:    c.UserEthAddress,
		EvmSignature:      c.EvmSignature,
		Amount:            c.Amount,
		UserCosmosAddress: c.UserCosmosAddress,
		Status:            c.Status,
		CosmosTxHash:      c.CosmosTxHash,
		EvmTxHash:         c.EvmTxHash,
		FailedReason:      c.FailedReason,
	}

}

func LikeCoinMigrationsFromModel(cs []*model.LikeCoinMigration) []*LikeCoinMigration {
	migrations := make([]*LikeCoinMigration, len(cs))
	for i, c := range cs {
		migrations[i] = LikeCoinMigrationFromModel(c)
	}
	return migrations
}
